### DISCUSSION TITLE
ORBAT chart edit mode

### DISCUSSION THREAD

**Discussion by kjellmf** [orbat-mapper/orbat-mapper#117] [Ideas]
An additional ORBAT editing mode based on the the ORBAT chart experiment. See #50 for an initial discussion. Should look like an ORBAT chart but interactive and editable.

**Comment by EzSalvucci** [comment]
> An additional ORBAT editing mode based on the the ORBAT chart experiment. See #50 for an initial discussion. Should look like an ORBAT chart but interactive and editable.

Hi, kjellmf. Your work until now is great! However, there is a problem: the ORBAT's Chart edit tool is not working. I don't know if you noticed it yet, but I'll tell you just in case.
Thank you for your work!
Sincerely,
Ez. Salvucci

**Comment by kjellmf** [comment]
@EzSalvucci Sorry about that. The chart edit mode button is currently just a placeholder. I have not started on that feature yet. 

I've done some work on this before (https://preview--orbat-mapper.netlify.app/chart), but I have not figured out yet the best way to  integrate it into the editor. If you have any ideas let me know.

Thank you for the feedback!

**Comment by EzSalvucci** [comment]
> @EzSalvucci Sorry about that. The chart edit mode button is currently just a placeholder. I have not started on that feature yet.
> 
> I've done some work on this before (https://preview--orbat-mapper.netlify.app/chart), but I have not figured out yet the best way to integrate it into the editor. If you have any ideas let me know.
> 
> Thank you for the feedback!

I don't have any idea: my programming knowledge is too shallow.

**Comment by kjellmf** [comment]
I've added a simple placeholder page when you click on the chart edit mode button (0ca030a).

**Comment by EzSalvucci** [comment]
I might have an idea on the way in wich can be done, but may do not understand the finer technical difficulties: I'm just an history  university student who appent to make out of the informatic technical high school.

However, i need to know what is the file you used for the ORBAT chart test...

P.S.: now also the grid edit mode as stopped working.

**Comment by kjellmf** [comment]
> I might have an idea on the way in wich can be done, but may do not understand the finer technical difficulties: I'm just an history university student who appent to make out of the informatic technical high school.

Sorry, I meant to say that if you have any ideas or wishes for for functionality/features let me know. The technical integration I'll handle myself.

> 
> However, i need to know what is the file you used for the ORBAT chart test...
> 

The chart is created from the Falklands war scenario.

**Comment by EzSalvucci** [comment]
Ok. So, do you know the site of http://www.niehorster.org/index.htm?

**Comment by kjellmf** [comment]
> Ok. So, do you know the site of http://www.niehorster.org/index.htm?

Thanks for reminding med about that site. It is a great resource! I should add a link to it on the landing page.

**Comment by EzSalvucci** [comment]
OK, basically my idea is this: any unit with more than 3 subordinate units layers should become a link to a sub-page with those layers, and, inside those pages, any sub-unit has more than 3 sub-subordinate layers should become a link to a sub-sub-page with more command layers, and so on...

Sorry for the ragged explanation... hope will help.

**Comment by EzSalvucci** [comment]
> > Ok. So, do you know the site of http://www.niehorster.org/index.htm?
> 
> Thanks for reminding med about that site. It is a great resource! I should add a link to it on the landing page.

How is going? I may have another idea, but can be superfluous...

**Comment by kjellmf** [comment]
This is a slowly progressing hobby project. I'm currently reviewing my old ORBAT chart code trying to remember how it works.

**Comment by EzSalvucci** [comment]
> This is a slowly progressing hobby project. I'm currently reviewing my old ORBAT chart code trying to remember how it works.

Ahhhh! My condolences... When you'll be finished, can you notify me? So my idea doesn't distract you...

**Comment by EzSalvucci** [comment]
Hi, kjellmf! Salvucci's here!

I noticed that you finally implemented the chart edit mode! Great Work!

However, I'm here to bring to your attention a bug in that! Sorry! :-(

The bug is this: 

1. if you create a unit, 
2. then create two or more subordinate units of that, 
3. and then create a subordinate for ONLY ONE of them,
4. then the sub-subordinate unit will not properly connect and align with the superior.

Probably you already noticed this bug, but I'll send this message just in case.

Thank you for your work and best regards,
E.Salvucci

**Reply by kjellmf** [reply]
@EzSalvucci The drawing of connectors is now better, but still not good enough. Need to alter the layout algorithm by drawing the bottom level first to determine the correct spacing.

**Reply by EzSalvucci** [reply]
A way to circumvent this need and implement an easier stopgap solution may be adding to the Unit Symbol customization menu a stacking option similar to the one of BattleOrder.
P.S.: I don't know if coding this is easier or harder than directly fixing the spacing...

**Comment by kjellmf** [comment]
I'm sorry, but I have not added any new functionality or fixed any bugs yet. I have just copied the ORBAT chart code into the the chart edit mode view as a staring point. The current layout algorithm is very simple.

**Comment by EzSalvucci** [comment]
Ahh! I understand.

Will help if I send you the ORBAT Chart I created with your tool to use as a debug tool?

Moreover, I have some suggestions to help with the chart edit mode and to make quality of life improvement to the creation of the ORBAT (even without the chart edit mode)

Thanks again for your work,
E. Salvucci

P.S.: don't feel pressed by my suggestions, you are doing a great job!
P.P.S.: do you have a Patreon?

**Reply by kjellmf** [reply]
Feel free to post a sample ORBAT. Nice to see what people are creating. 
I'm happy for all suggestions. I can't promise I'll implement them, but making the ORBAT creation process as smooth as possible is my goal. 

Thanks for the feedback!

**Reply by kjellmf** [reply]
I tried recently to recreate some of the ORBATs from https://www.battleorder.org/us-orbat. I identified several pain points in the ORBAT making process that I will try to fix.

**Reply by EzSalvucci** [reply]
Umm, how i do post it? It seems that GitHub doesn't support KML, KMZ or GeoJSON files...

**Reply by kjellmf** [reply]
For short snippets of code you can embed it as a code block i a comment. You can also create a github gist and share a link here https://gist.github.com/

**Reply by EzSalvucci** [reply]
is this okay?

https://gist.github.com/EzSalvucci/6adf8919cbc3136e80151895356f97ad

**Reply by kjellmf** [reply]
Nice! Thanks for sharing.

**Reply by EzSalvucci** [reply]
Thank you! 

So, about my suggestions.

I suggest:
1. make so that the "Duplicate" option will fully duplicate a unit, including her subordinate units;
    a. or add a "Fully Duplicate" option that will do that
2. make so that using the "Add Root Unit" and "Add Subordinate" options will not automatically generate a unit; 
    a. instead using them will open a menù in which you can control the details of the unit you are about to generate;
    b. maybe with some preselected options (like the unit size being smaller than the parent unit for the Add Subordinate option);
3. add a stacking option like that in the https://www.battleorder.org/icons

These are my suggestion that I think will improve the ease of use of your ORBAT Mapper tool. Hope that are useful.

Thank you for your work,
E. Salvucci

P.S.: sorry if ask this again, but do you have a Patreon?

**Reply by kjellmf** [reply]
Thank you for the suggestions. I'll keep them in mind. Suggestion 1 and 3 is already on my TODO list. Suggestion 2 is really a good idea. 

Thank you for asking, but I don't have a Patreon (or any other similar service). My main motivation is learning and having fun.

**Reply by kjellmf** [reply]
Suggestion 1 is now implemented (https://github.com/orbat-mapper/orbat-mapper/commit/5420e00ef25a484cd51ee2de20b7aa905badde32)

**Comment by kjellmf** [comment]
I've converted the issue to a discussion for easier exchange of ideas.
